chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,437 @@
+        n_instances = gt.shape[1]
+
+        p = expand_dims(pred, axis=1)
+        g = expand_dims(gt, axis=2)
+        m = expand_dims(means, axis=3)
+
+        p = broadcast_to(p, (bs, n_instances, n_filters, n_loc))
+        g = broadcast_to(g, (bs, n_instances, n_filters, n_loc))
+        m = broadcast_to(m, (bs, n_instances, n_filters, n_loc))
+
+        m = cast(m, p.dtype)
+        g = cast(g, p.dtype)
+
+        xp = cuda.get_array_module(p)
+        dv = xp.asarray(delta_v, p.dtype)
+
+        _var = self.norm((p - m), 2)
+        _var = maximum(xp.asarray(0.0, p.dtype),
+                       _var - dv) ** 2  # Suppress inlier distance
+        _var = _var * g[:, :, 0, :]
+
+        var_term = 0.0
+        for i in range(bs):
+            if len(gt_idx[i]) < 1:
+                warnings.warn("Warning : Empty gt_idx is found. " +
+                              "Please check dataset content.")
+                continue
+
+            gt_sm = c_sum(g[i, gt_idx[i], 0])
+            if gt_sm.data == 0:
+                warnings.warn("Warning : Zero gt_sm is found. " +
+                              "Please check dataset content.")
+                continue
+
+            var_sm = c_sum(_var[i, gt_idx[i]])
+            if var_sm.data == 0:
+                continue
+
+            var_term += (var_sm / gt_sm)
+        var_term /= bs
+
+        return var_term
+
+    def _distance_term(self, means, delta_d, n_objects):
+        """Function to calculate distance term
+
+        Args:
+            means (nd-array): Instance means
+            delta_d (float): Coefficient to decide 'push force' power
+            n_objects (nd-array): Instance count in current input
+
+        Returns:
+            float : distance loss
+        """
+
+        bs, n_instances, n_filters = means.shape
+        m = cast(means, means.dtype)
+
+        module = cuda.get_array_module(means)
+        dd = module.asarray(delta_d, means.dtype)
+
+        dist_term = 0.0
+        for i in range(bs):
+
+            if n_objects[i] <= 1:
+                continue
+
+            nobj = n_objects[i]
+
+            # Prepare means
+            m_i = expand_dims(m[i, :nobj, :], 1)
+            m_1 = broadcast_to(m_i, (nobj, nobj, n_filters))
+            m_2 = transpose(m_1, axes=(1, 0, 2))
+
+            nrm = self.norm(m_1 - m_2, axis=2)
+            margin = 2.0 * dd * (1.0 - module.eye(nobj, dtype=means.dtype))
+
+            _dist_term_sample = c_sum(
+                maximum(module.asarray(0.0, means.dtype), margin - nrm) ** 2)
+            _dist_term_sample /= nobj * (nobj - 1)
+            dist_term += _dist_term_sample
+
+        dist_term /= bs
+
+        return dist_term
+
+    def _regularization_term(self, means, gt_idx):
+        """Function to calculate regularization term
+
+        Args:
+            means (nd-array): Instance means
+            gt_idx (tuple / nd-array): Indexes of ground truth instances
+
+        Returns:
+            float : distance loss
```

Comment:
Please indicate that the return value is a `chainer.Variable`, not a float.

Reply:
@dBeker Could you reflect @tkerola 's comment?